Handle flaky upload size test

We have very often following flaky test:

**com.daimler.sechub.integrationtest.scenario2.FileUploadSizeScenario2IntTest.when_file_exceeds_5MB_a_server_error_is_thrown**

Here we expect an `HttpServerErrorException` but some times we got a `SocketException` (see following example stacktrace ) what is really disturbing because we often got broken builds where everything is okay.

We have following possibilities
- drop this test and replace by check for correct spring boot setup (so 5MB is upload size)
- drop this test and just dont test this
- examine why we got here an `SocketException` and fix  root cause (if possible, I got the feeling this has something to do with a race condition of throwing exception an closing server socket...)
- ... ?

```stacktrace
java.lang.AssertionError: 
Expected: an instance of org.springframework.web.client.HttpServerErrorException
     but: <org.springframework.web.client.ResourceAccessException: I/O error on POST request for "https://localhost:8243/api/project/scenario2_project1/job/04666978-a301-4fd6-b680-112acbd91804/sourcecode": Connection reset; nested exception is java.net.SocketException: Connection reset> is a org.springframework.web.client.ResourceAccessException
Stacktrace was: org.springframework.web.client.ResourceAccessException: I/O error on POST request for "https://localhost:8243/api/project/scenario2_project1/job/04666978-a301-4fd6-b680-112acbd91804/sourcecode": Connection reset; nested exception is java.net.SocketException: Connection reset
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:751)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:677)
	at org.springframework.web.client.RestTemplate.postForEntity(RestTemplate.java:452)
	at com.daimler.sechub.integrationtest.internal.TestRestHelper.upload(TestRestHelper.java:202)
	at com.daimler.sechub.integrationtest.api.AsUser.upload(AsUser.java:59)
	at com.daimler.sechub.integrationtest.scenario2.FileUploadSizeScenario2IntTest.handleBigUpload(FileUploadSizeScenario2IntTest.java:93)
	at com.daimler.sechub.integrationtest.scenario2.FileUploadSizeScenario2IntTest.when_file_exceeds_5MB_a_server_error_is_thrown(FileUploadSizeScenario2IntTest.java:46)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at com.daimler.sechub.integrationtest.api.IntegrationTestSetup$IntegrationTestStatement.evaluate(IntegrationTestSetup.java:159)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketException: Connection reset
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at sun.security.ssl.OutputRecord.writeBuffer(OutputRecord.java:431)
	at sun.security.ssl.OutputRecord.write(OutputRecord.java:417)
	at sun.security.ssl.SSLSocketImpl.writeRecordInternal(SSLSocketImpl.java:879)
	at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:850)
	at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:123)
	at org.apache.http.impl.io.SessionOutputBufferImpl.streamWrite(SessionOutputBufferImpl.java:124)
	at org.apache.http.impl.io.SessionOutputBufferImpl.write(SessionOutputBufferImpl.java:160)
	at org.apache.http.impl.io.ContentLengthOutputStream.write(ContentLengthOutputStream.java:113)
	at org.apache.http.entity.ByteArrayEntity.writeTo(ByteArrayEntity.java:112)
	at org.apache.http.impl.DefaultBHttpClientConnection.sendRequestEntity(DefaultBHttpClientConnection.java:156)
	at org.apache.http.impl.conn.CPoolProxy.sendRequestEntity(CPoolProxy.java:152)
	at org.apache.http.protocol.HttpRequestExecutor.doSendRequest(HttpRequestExecutor.java:238)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:123)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at org.springframework.http.client.HttpComponentsClientHttpRequest.executeInternal(HttpComponentsClientHttpRequest.java:87)
	at org.springframework.http.client.AbstractBufferingClientHttpRequest.executeInternal(AbstractBufferingClientHttpRequest.java:48)
	at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpRequest.java:53)
	at org.springframework.http.client.BufferingClientHttpRequestWrapper.executeInternal(BufferingClientHttpRequestWrapper.java:63)
	at org.springframework.http.client.AbstractBufferingClientHttpRequest.executeInternal(AbstractBufferingClientHttpRequest.java:48)
	at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpRequest.java:53)
	at org.springframework.http.client.InterceptingClientHttpRequest$InterceptingRequestExecution.execute(InterceptingClientHttpRequest.java:109)
	at com.daimler.sechub.integrationtest.internal.TestSecHubRestAPIClientHttpRequestInterceptor.intercept(TestSecHubRestAPIClientHttpRequestInterceptor.java:52)
	at org.springframework.http.client.InterceptingClientHttpRequest$InterceptingRequestExecution.execute(InterceptingClientHttpRequest.java:93)
	at org.springframework.http.client.InterceptingClientHttpRequest.executeInternal(InterceptingClientHttpRequest.java:77)
	at org.springframework.http.client.AbstractBufferingClientHttpRequest.executeInternal(AbstractBufferingClientHttpRequest.java:48)
	at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpRequest.java:53)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:742)
	... 20 more

```

--- a/sechub-integrationtest/src/test/java/com/daimler/sechub/integrationtest/scenario2/FileUploadSizeScenario2IntTest.java
+++ b/sechub-integrationtest/src/test/java/com/daimler/sechub/integrationtest/scenario2/FileUploadSizeScenario2IntTest.java
@@ -17,7 +17,7 @@ import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
 import org.junit.rules.Timeout;
-import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.HttpClientErrorException.NotAcceptable;
 
 import com.daimler.sechub.integrationtest.api.IntegrationTestSetup;
 import com.daimler.sechub.sharedkernel.util.FileChecksumSHA256Service;
@@ -41,7 +41,7 @@ public class FileUploadSizeScenario2IntTest {
 	 * @throws IOException
 	 */
 	@Test
-	public void when_file_exceeds_5MB_a_server_error_is_thrown() throws IOException {
+	public void when_file_exceeds_5MB_a_NOT_ACCEPTABLE_is_returned() throws IOException {
 		/* @formatter:off */
 		handleBigUpload(true);
 	}
@@ -72,18 +72,7 @@ public class FileUploadSizeScenario2IntTest {
 
 		/* test */
 		if (tooBig) {
-//			We do not expect ResourceAccessException here but a 500, see
-//			https://stackoverflow.com/questions/48891490/sizelimitexceededexception-not-being-caught-by-spring-controlleradvice
-//			https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/multipart/support/MultipartFilter.html
-//			Did try suggestions done at https://www.baeldung.com/exception-handling-for-rest-with-spring but did not work at all
-//			we got always
-//
-//			org.apache.tomcat.util.http.fileupload.FileUploadBase$SizeLimitExceededException: the request was rejected because its size (5244303) exceeds the configured maximum (5242880)
-//			at org.apache.tomcat.util.http.fileupload.FileUploadBase$FileItemIteratorImpl.<init>(FileUploadBase.java:802) ~[tomcat-embed-core-8.5.32.jar:8.5.32]
-//          done by apache server which leads to a 500.
-//
-//			would be nice to to throw a NotAcceptableException instead
-			expected.expect(HttpServerErrorException.class);
+			expected.expect(NotAcceptable.class);
 		}else {
 			/* nothing - means expected no exception at all!*/
 		}
@@ -100,13 +89,13 @@ public class FileUploadSizeScenario2IntTest {
 	 * filename, sha256checksum,..)
 	 */
 	private File createZipFileContainingMegabytes(boolean uploadShallBeTooLarge) throws FileNotFoundException, IOException {
-		String tmpOutputFilePath = "build/resources/bigFile";
+		String tmpPath = "build/resources/bigFile";
 		if (uploadShallBeTooLarge) {
-			tmpOutputFilePath += "-too-large";
+		    tmpPath += "-too-large";
 		} else {
-			tmpOutputFilePath += "-accepted";
+		    tmpPath += "-accepted";
 		}
-		tmpOutputFilePath += ".zip";
+		String tmpZipFilePath = tmpPath+ ".zip";
 
 		int maximumUploadSizeInMB = 5;
 		int maximumUploadSizeInBytes = 1024 * 1024 * maximumUploadSizeInMB;
@@ -116,18 +105,17 @@ public class FileUploadSizeScenario2IntTest {
 														// checksum on upload)
 		}
 		byte[] content = new byte[bytesToOrder];
-
-		try (FileOutputStream fileOutputStream = new FileOutputStream(tmpOutputFilePath);
+		try (FileOutputStream fileOutputStream = new FileOutputStream(tmpZipFilePath);
 				ZipOutputStream zipOutputStream = new ZipOutputStream(new BufferedOutputStream(fileOutputStream));) {
 
-			ZipEntry zipEntry = new ZipEntry(tmpOutputFilePath);
+			ZipEntry zipEntry = new ZipEntry("test.bin");
 			// Set compression level to minimum to generate big zip file
 			zipOutputStream.setLevel(0);
 			zipOutputStream.putNextEntry(zipEntry);
 			zipOutputStream.write(content);
 			zipOutputStream.flush();
 		}
-		File file = new File(tmpOutputFilePath);
+		File file = new File(tmpZipFilePath);
 		if (uploadShallBeTooLarge && file.length() < maximumUploadSizeInBytes) {
 			throw new IllegalStateException("Wanted at least file size: " + maximumUploadSizeInBytes + " but was:" + file.length());
 		}

--- a/sechub-schedule/src/main/java/com/daimler/sechub/domain/schedule/SchedulerUploadService.java
+++ b/sechub-schedule/src/main/java/com/daimler/sechub/domain/schedule/SchedulerUploadService.java
@@ -57,7 +57,7 @@ public class SchedulerUploadService {
 
 	@Autowired
 	UserInputAssertion assertion;
-
+	
 	@UseCaseUserUploadsSourceCode(@Step(number = 2, name = "Try to find project annd upload sourcecode as zipfile", description = "When project is found and user has access and job is initializing the sourcecode file will be uploaded"))
 	public void uploadSourceCode(String projectId, UUID jobUUID, MultipartFile file, String checkSum) {
 		assertion.isValidProjectId(projectId);

--- a/sechub-server/src/main/java/com/daimler/sechub/server/SecHubExceptionHandler.java
+++ b/sechub-server/src/main/java/com/daimler/sechub/server/SecHubExceptionHandler.java
@@ -1,0 +1,40 @@
+package com.daimler.sechub.server;
+
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.tomcat.util.http.fileupload.FileUploadBase.SizeLimitExceededException;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.multipart.MaxUploadSizeExceededException;
+
+@ControllerAdvice
+public class SecHubExceptionHandler {
+
+    @ExceptionHandler(SizeLimitExceededException.class)
+    @ResponseBody
+    public Exception handleFileUploadSizeExceeded(SizeLimitExceededException ex, HttpServletResponse response) {
+        return commonHandleFileUploadSizeExceed(ex, response);
+    }
+
+    @ExceptionHandler(MaxUploadSizeExceededException.class)
+    @ResponseBody
+    public Exception handleFileUploadSizeExceeded(MaxUploadSizeExceededException ex, HttpServletResponse response) {
+        return commonHandleFileUploadSizeExceed(ex, response);
+    
+    }
+    
+    private Exception commonHandleFileUploadSizeExceed(Exception ex, HttpServletResponse response) {
+        if (response == null) {
+            throw new IllegalStateException("response missing");
+        }
+        if (ex == null) {
+            throw new IllegalStateException("exception missing");
+        }
+        
+        response.setStatus(HttpStatus.NOT_ACCEPTABLE.value());
+        
+        return ex;
+    }
+}

--- a/sechub-server/src/main/java/com/daimler/sechub/server/SecHubTomcatServletWebserFactory.java
+++ b/sechub-server/src/main/java/com/daimler/sechub/server/SecHubTomcatServletWebserFactory.java
@@ -1,0 +1,51 @@
+package com.daimler.sechub.server;
+
+import org.apache.catalina.connector.Connector;
+import org.apache.coyote.http11.AbstractHttp11Protocol;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.web.embedded.tomcat.TomcatConnectorCustomizer;
+import org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory;
+import org.springframework.boot.web.server.WebServerFactoryCustomizer;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SecHubTomcatServletWebserFactory implements WebServerFactoryCustomizer<TomcatServletWebServerFactory> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SecHubTomcatServletWebserFactory.class);
+
+    @Value("${spring.servlet.multipart.max-file-size}")
+    private String maxFileSizeAsString;
+
+    @Override
+    public void customize(TomcatServletWebServerFactory factory) {
+        /* @formatter:off 
+          
+           we let Tomcat always swallow one more megabyte than spring servlet filter does
+           so we will get not SocketConnectionExceptions sometimes instad dedicated exceptions
+           as defined in SecHubExceptionHandler 
+           
+           @formatter:on*/
+        String size = maxFileSizeAsString.trim().toLowerCase();
+        if (!size.endsWith("mb")) {
+            LOG.error("Cannot customize tomcat swallow size automatically, because multipart size not supported:" + size);
+            return;
+        }
+        size = size.substring(0, size.length() - 2);
+        size = size.trim();
+
+        int springUploadMaxFileSizeInMB = Integer.parseInt(size);
+        int maxSwallowSize = (springUploadMaxFileSizeInMB + 1) * 1024 * 1024;
+
+        factory.addConnectorCustomizers(new TomcatConnectorCustomizer() {
+            @Override
+            public void customize(Connector connector) {
+                if (connector.getProtocolHandler() instanceof AbstractHttp11Protocol) {
+                    LOG.info("Set max swallow size to {}", maxSwallowSize);
+                    ((AbstractHttp11Protocol<?>) connector.getProtocolHandler()).setMaxSwallowSize(maxSwallowSize);
+                }
+            }
+        });
+    }
+}

